### PR TITLE
implement new database config persistence

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/8439

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt
Comment (subodh1810) [inline airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt:1]: We use the following piece of code to create the dump 
```
ExceptionWrappingDatabase(database).query(ctx -> ctx.meta().ddl().queryStream()
        .map(query -> query.toString() + ";")
        .filter(statement -> !statement.startsWith("create schema"))
        .collect(Collectors.joining("\n")));
```
and for some reason it doesnt acknowledge the enum types. I had added these manually but these keep getting deleted. AM trying to figure out how to stop it from deleting these

File: airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/BaseDatabaseConfigPersistence2Test.java
Comment (subodh1810) [inline airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/BaseDatabaseConfigPersistence2Test.java:36]: The test classes `BaseDatabaseConfigPersistence2Test`, `DatabaseConfigPersistence2LoadDataTest` , `DatabaseConfigPersistence2MigrateFileConfigsTest`, `DatabaseConfigPersistence2Test`  are inspired by the existing tests `BaseDatabaseConfigPersistenceTest`, `DatabaseConfigPersistenceLoadDataTest`, `DatabaseConfigPersistenceMigrateFileConfigsTest` , `DatabaseConfigPersistenceTest`

File: airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt
Comment (cgardens) [inline airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt:9]: i'm assuming this is just noise?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:115]: why do we need to be able to switch back to the old one?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:376]: nice reuse

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:122]: i wonder if you can save your self some duplication by piggybacking off of the list queries. e.g. getStandardWorkspaceWithMetadata() optionally took in a config id or you split up `getStandardWorkspaceWithMetadata` to expose the underlying query? there a few ways you might be able to do it.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:1267]: is it possible to dedupe all of the deletions. fundamentally they are all doing `DELETE FROM <table name> WHERE ID=?`. I know there are one or 2 exceptions but with this approach i think you can remove a lot of lines of code.

**Review by cgardens — COMMENTED** [review]
Looks good! I didn't have a chance to read the whole thing yet which is why I did not approve yet. That said, the direction looks very good to me. Figured I would share comments now to start discussion as early as possible. Will finish reviewing tomorrow.

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:45]: can we use a more explicit name here?

Comment (michel-tricot) [reply airbyte-db/lib/src/main/resources/configs_database/schema2.sql:45]: `send_newsletter`?

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:46]: `send_security_updates`

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:50]: `first_sync_complete` (or `first_sync_completed`, if you go for this one then the column above should be `initial_setup_completed`)

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:51]: `feedback_completed`

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:71]: this field shouldn't belong here since it is only for sources

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:11]: @cgardens should we start using the official v2 names?

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:156]: why is the casing different?

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (michel-tricot) [inline airbyte-db/lib/src/main/resources/configs_database/schema2.sql:37]: we should be careful about creating consistency in columns names. that will be use as an example in the future.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (davinchia) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:88]: why is this always set to false?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (davinchia) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:374]: should all these methods start with `list` instead of `get`? my initial reaction was that these were retrieving one record instead of all the records like it's doing.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (davinchia) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:1267]: +1

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (davinchia) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java:78]: This large class makes me think it's probably times for each database model to have it's own persistence class instead of having a single class responsible for all this. We don't have to do this now to avoid changing too many things at once, but I'm thinking we should inset a comment here that says we should break away from this pattern moving forward and slowly migrate the older tables over. 

How does this sound? @cgardens ?

**Review by davinchia — COMMENTED** [review]
Generally made sense to me.

Minor comments on naming/clarity. One discussion comment I want to resolve before we merge this in.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java
Comment (subodh1810) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java:17]: Unused constructor, was confusing regarding why it is accepting the schema_dump file

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/JobsFlywayMigrationDatabase.java
Comment (subodh1810) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/JobsFlywayMigrationDatabase.java:17]: same as above

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/test/TestDatabaseProviders.java
Comment (subodh1810) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/test/TestDatabaseProviders.java:27]: Unused constructor

File: airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt
Comment (subodh1810) [reply airbyte-db/lib/src/main/resources/configs_database/schema_dump.txt:1]: the schema_dump file is just for representation purpose and isnt used to generate or initialize anything. added a comment to explain it

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/FlywayDatabaseMigrator.java
Comment (subodh1810) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/FlywayDatabaseMigrator.java:55]: unused method

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java
Comment (subodh1810) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java]: @davinchia as discussed over call, thats the plan. For now we are just changing the DatabaseConfigPersistence so that we dont have to change any application level code. As part of leveraging the new table structure we would head into that direction

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsDatabaseInstance.java
Comment (davinchia) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsDatabaseInstance.java:55]: can we leave a quick comment to explain why we specifically check for the 'state' table. thanks

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsDatabaseTables.java
Comment (davinchia) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsDatabaseTables.java:14]: sort

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/DeprecatedConfigsDatabaseTables.java
Comment (davinchia) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/DeprecatedConfigsDatabaseTables.java:11]: should we add @Deprecated here?

**Review by davinchia — APPROVED** [review]
Nice!

Appreciate removing the unused stuff and the added comments.

Minor comments about adding the deprecated annotations. Non-blocking.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java
Comment (cgardens) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java:17]: this isn't used by flyway somehow? if i'm wrong ignore it. @tuliren

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (cgardens) [reply airbyte-db/lib/src/main/resources/configs_database/schema2.sql]: agreed. this is a conscious choice that @subodh1810 and I made. This is the only difference between Source and Destination Definition right now. As they drift further apart we will do a proper polymorphic association. In the meanwhile adding that complexity for this field felt like the wrong tradeoff. 

@subodh1810 we should add comment (wherever possible, i know it's probably not possible in this dump file) explaining this tradoeff.

File: airbyte-db/lib/src/main/resources/configs_database/schema2.sql
Comment (cgardens) [reply airbyte-db/lib/src/main/resources/configs_database/schema2.sql]: i think we should handle it when we work on that protocol v2 project. until then these enum names are more clear and correct when describing what's actually there.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java
Comment (tuliren) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/ConfigsFlywayMigrationDatabase.java:17]: Yes, confirmed that this constructor is no longer used.

File: airbyte-db/lib/src/main/resources/configs_database/normalized_tables_schema.txt
Comment (cgardens) [inline airbyte-db/lib/src/main/resources/configs_database/normalized_tables_schema.txt:1]: this is pretty.

**Pull request merged**
